File: test_match_publisher.py
#!/usr/bin/env python3
"""
Script to publish test match.ended events to RabbitMQ for testing geo-data-service
without needing to play actual games.
"""

import pika
import json
import uuid
from datetime import datetime, timezone

# RabbitMQ Configuration
RABBITMQ_HOST = "location404-rabbitmq-8b2418-181-215-135-221.traefik.me"
RABBITMQ_PORT = 5672
RABBITMQ_USER = "admin"
RABBITMQ_PASSWORD = "rqtisgqs"
RABBITMQ_VHOST = "/"
EXCHANGE_NAME = "game-events"
ROUTING_KEY = "match.ended"

def create_test_match_event(player_a_id: str, player_b_id: str):
    """
    Creates a test GameMatchEndedEventDto
    """
    match_id = str(uuid.uuid4())
    now = datetime.now(timezone.utc).isoformat()

    # Create 3 rounds with realistic data
    rounds = []
    for i in range(1, 4):
        round_id = str(uuid.uuid4())

        # Random coordinates for correct answer
        correct_x = -23.5505 + (i * 0.01)  # Near São Paulo
        correct_y = -46.6333 + (i * 0.01)

        # Player A guess (closer to correct answer)
        player_a_x = correct_x + 0.005
        player_a_y = correct_y + 0.005

        # Player B guess (farther from correct answer)
        player_b_x = correct_x + 0.02
        player_b_y = correct_y + 0.02

        rounds.append({
            "id": round_id,
            "roundNumber": i,
            "gameResponse": {
                "x": correct_x,
                "y": correct_y
            },
            "playerAGuess": {
                "x": player_a_x,
                "y": player_a_y
            },
            "playerBGuess": {
                "x": player_b_x,
                "y": player_b_y
            }
        })

    event = {
        "matchId": match_id,
        "playerAId": player_a_id,
        "playerBId": player_b_id,
        "rounds": rounds,
        "endedAt": now
    }

    return event

def publish_match_event(event):
    """
    Publishes a match.ended event to RabbitMQ
    """
    credentials = pika.PlainCredentials(RABBITMQ_USER, RABBITMQ_PASSWORD)
    parameters = pika.ConnectionParameters(
        host=RABBITMQ_HOST,
        port=RABBITMQ_PORT,
        virtual_host=RABBITMQ_VHOST,
        credentials=credentials
    )

    connection = pika.BlockingConnection(parameters)
    channel = connection.channel()

    # Ensure exchange exists
    channel.exchange_declare(
        exchange=EXCHANGE_NAME,
        exchange_type='topic',
        durable=True
    )

    # Publish message
    message = json.dumps(event)
    channel.basic_publish(
        exchange=EXCHANGE_NAME,
        routing_key=ROUTING_KEY,
        body=message,
        properties=pika.BasicProperties(
            delivery_mode=2,  # Make message persistent
            content_type='application/json'
        )
    )

    print(f"✓ Published match.ended event:")
    print(f"  Match ID: {event['matchId']}")
    print(f"  Player A: {event['playerAId']}")
    print(f"  Player B: {event['playerBId']}")
    print(f"  Rounds: {len(event['rounds'])}")

    connection.close()

def main():
    # Use test player IDs - you can replace these with actual player IDs from your database
    # Or pass them as command line arguments
    player_a_id = "e19cfe2c-4f56-4914-841b-0aac8b46be24"  # From the match we verified
    player_b_id = "0e3cc492-ec0f-4fc3-a2de-dfa2f84b6fa3"  # From the match we verified

    print("Creating test match event...")
    event = create_test_match_event(player_a_id, player_b_id)

    print("Publishing to RabbitMQ...")
    publish_match_event(event)

    print("\n✓ Done! Check geo-data-service logs to verify processing.")
    print(f"  You can query the match at: http://localhost:5000/api/matches/{event['matchId']}")

if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"✗ Error: {e}")
        import traceback
        traceback.print_exc()

File: tests/scripts/test_match_publisher.lua
local http = require("socket.http")
local ltn12 = require("ltn12")
local json = require("cjson")

local RABBITMQ_HOST = ""
local RABBITMQ_PORT = "15672"  -- Porta HTTP API
local RABBITMQ_USER = "admin"
local RABBITMQ_PASSWORD = ""
local EXCHANGE_NAME = "game-events"
local ROUTING_KEY = "match.ended"

local PLAYER_A_ID = "e19cfe2c-4f56-4914-841b-0aac8b46be24"
local PLAYER_B_ID = "0e3cc492-ec0f-4fc3-a2de-dfa2f84b6fa3"

-- Função para gerar UUID simples
local function generate_uuid()
    local template = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'
    return string.gsub(template, '[xy]', function (c)
        local v = (c == 'x') and math.random(0, 0xf) or math.random(8, 0xb)
        return string.format('%x', v)
    end)
end

-- Função para obter timestamp ISO 8601
local function get_iso_timestamp()
    return os.date("!%Y-%m-%dT%H:%M:%S.000Z")
end

-- Função para criar evento de teste
local function create_test_match_event()
    local match_id = generate_uuid()
    local timestamp = get_iso_timestamp()

    local event = {
        matchId = match_id,
        playerAId = PLAYER_A_ID,
        playerBId = PLAYER_B_ID,
        rounds = {
            {
                id = generate_uuid(),
                roundNumber = 1,
                gameResponse = {
                    x = -23.5505,
                    y = -46.6333
                },
                playerAGuess = {
                    x = -23.5455,
                    y = -46.6283
                },
                playerBGuess = {
                    x = -23.5305,
                    y = -46.6133
                }
            },
            {
                id = generate_uuid(),
                roundNumber = 2,
                gameResponse = {
                    x = -23.5605,
                    y = -46.6433
                },
                playerAGuess = {
                    x = -23.5555,
                    y = -46.6383
                },
                playerBGuess = {
                    x = -23.5405,
                    y = -46.6233
                }
            },
            {
                id = generate_uuid(),
                roundNumber = 3,
                gameResponse = {
                    x = -23.5705,
                    y = -46.6533
                },
                playerAGuess = {
                    x = -23.5655,
                    y = -46.6483
                },
                playerBGuess = {
                    x = -23.5505,
                    y = -46.6333
                }
            }
        },
        endedAt = timestamp
    }

    return event
end

-- Função para publicar evento usando HTTP API do RabbitMQ
local function publish_match_event(event)
    local url = string.format(
        "http://%s:%s@%s:%s/api/exchanges/%%2F/%s/publish",
        RABBITMQ_USER,
        RABBITMQ_PASSWORD,
        RABBITMQ_HOST,
        RABBITMQ_PORT,
        EXCHANGE_NAME
    )

    local payload = {
        properties = {
            delivery_mode = 2,
            content_type = "application/json"
        },
        routing_key = ROUTING_KEY,
        payload = json.encode(event),
        payload_encoding = "string"
    }

    local request_body = json.encode(payload)
    local response_body = {}

    local res, code, headers, status = http.request{
        url = url,
        method = "POST",
        headers = {
            ["Content-Type"] = "application/json",
            ["Content-Length"] = tostring(#request_body)
        },
        source = ltn12.source.string(request_body),
        sink = ltn12.sink.table(response_body)
    }

    if code == 200 then
        local response = json.decode(table.concat(response_body))
        if response.routed then
            print("✓ Evento publicado com sucesso!")
            print(string.format("  Match ID: %s", event.matchId))
            print(string.format("  Player A: %s", event.playerAId))
            print(string.format("  Player B: %s", event.playerBId))
            print(string.format("  Rounds: %d", #event.rounds))
            print()
            print("Verifique os logs do geo-data-service para confirmar o processamento.")
            print(string.format("Consulte a partida em: http://localhost:5000/api/matches/%s", event.matchId))
            return true
        else
            print("✗ Mensagem não foi roteada")
            return false
        end
    else
        print(string.format("✗ Erro ao publicar: HTTP %d", code))
        print(table.concat(response_body))
        return false
    end
end

-- Main
math.randomseed(os.time())

print("Criando evento de teste...")
local event = create_test_match_event()

print("Publicando no RabbitMQ...")
local success = publish_match_event(event)

if not success then
    os.exit(1)
end
